proposals/20210526-model_load_test_component.md

```diff
@@ -0,0 +1,63 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+**Your name:** Michal Brys
+
+**Your email:** michal.brys@openx.com
+
+**Your company/organization:** OpenX
+
+**Project name:** Model load test TFX Component
+
+## Project Description
+A TFX component performs a load test of the exported TensorFlow model because the prediction time may vary on the model type and structure.
+
+## Project Category
+Component
+
+## Project Use-Case(s)
+This component could extend a TFX InfraValidator behavior to validate the TensorFlow model serving performance using the load test.
+The load test results will tell us:
+* if the model will meet the prediction time requirements (essential in large scale implementations),
+* if the changes in the new model will make it faster/slower comparing to the baseline model.
+
+## Project Implementation
+The load test component will integrate with TFX using the following approach:
+* The `Model Load Test TFX Pipeline Component` creates a TensorFlow Serving endpoint like in the `InfraValidator TFX Pipeline Component`.
+* The load test component spins up a pod on the Kubernetes cluster to perform the load test (send a specified volume of the prediction requests to the TensorFlow Serving endpoint).
+* The model serving performance metrics (i.e. average prediction time) and the load test specification (the traffic volume, the network protocol used, etc.) are recorded and stored as an ML Metadata artifact.
+* Produced ML Metadata artifacts could validate prediction speed compared to the baseline model. Based on that, we can decide if we want to proceed or break the TFX pipeline execution.
+
+### Proposed component spec
+
+INPUTS:
+* `examples: ChannelParameter(type=tfx.types.standard_artifacts.Examples)` - examples used to perform load test. For the reproducibility reason, they should be the same across the multiple load tests, i.e. imported by `tfx.dsl.components.common.importer.Importer` component.
+* `model: ChannelParameter(type=tfx.types.standard_artifacts.Model)` - model used to perform load test. May be output of `Trainer` or `tfx.dsl.components.common.importer.Importer`.
+* `baseline_load_test_results:  ChannelParameter(type=ModelLoadTestResult`) - (optional) the artifact with the load test of the baseline model result. We will use it to compare models speed in the relative way.
```

Comment:
Let's define what `ModelLoadTestResult` artifact is. Typically artifact have a directory with certain structure (for example fixed filename), and artifact properties (map of str to {number|str|ProtoStruct} values).

My recommendation is to put load test result into a single JSON field artifact property, and leave the directory (artifact.uri) empty unless you have any file-type artifacts:

```python
# I chose the different name here...
class ModelPerformanceMetric(Artifact):
  TYPE_NAME = "ModelPerformanceMetric"
  PROPERTIES = {
    "perf_metric": Property(type=PropertyType.JSON),
    "blessed": Property(type=PropertyType.INT, optional=True)
  }
```

Where perf_metric corresponds to a JSON something like..

```json
{
  "settings": {
    "machine_type": "n1-standard-16",
    "rps": 100
  },
  "result": {
    "inference_time_ms": {
      "average": 14,
      "p50": 16,
      "p75": 22,
      "p90": 29,
      "p99": 44
    }
  }
}
```